proposed/2022/dotnet-nuget-why-proposal.md

```diff
@@ -0,0 +1,222 @@
+
+Sample `project.assets.json` file:
+
+```
+{
+  "version": 3,
+  "targets": {
+    "net6.0": {
+      "Microsoft.CSharp/4.3.0": {
+        "type": "package",
+        "dependencies": {
+          "System.Collections": "4.3.0",
+          "System.Diagnostics.Debug": "4.3.0",
+          "System.Dynamic.Runtime": "4.3.0",
+          "System.Globalization": "4.3.0",
+          "System.Linq": "4.3.0",
+          "System.Linq.Expressions": "4.3.0",
+          "System.ObjectModel": "4.3.0",
+          "System.Reflection": "4.3.0",
+          "System.Reflection.Extensions": "4.3.0",
+          "System.Reflection.Primitives": "4.3.0",
+          "System.Reflection.TypeExtensions": "4.3.0",
+          "System.Resources.ResourceManager": "4.3.0",
+          "System.Runtime": "4.3.0",
+          "System.Runtime.Extensions": "4.3.0",
+          "System.Runtime.InteropServices": "4.3.0",
+          "System.Threading": "4.3.0"
+        },
+        "compile": {
+          ...
+          }
+        },
+        "runtime": {
+          ...
+        }
+      },
+      "Microsoft.ML/1.7.1": {
+        "type": "package",
+        "dependencies": {
+          "Microsoft.ML.CpuMath": "1.7.1",
+          "Microsoft.ML.DataView": "1.7.1",
+          "Newtonsoft.Json": "10.0.3",
+          "System.CodeDom": "4.4.0",
+          "System.Collections.Immutable": "1.5.0",
+          "System.Memory": "4.5.3",
+          "System.Reflection.Emit.Lightweight": "4.3.0",
+          "System.Threading.Channels": "4.7.1"
+        },
+        "compile": {
+          ...
+        },
+        "runtime": {
+          ...
+        },
+        "build": {
+          ...
+        },
+        "runtimeTargets": {
+          ...
+      },
+  },
+  "libraries": {
+    ...
+  },
+  "projectFileDependencyGroups": {
+    "net6.0": [
+      "Microsoft.ML >= 1.7.1",
+      "Microsoft.ML.SampleUtils >= 0.19.1"
+    ]
+  },
+  "packageFolders": {
+    ...
+  },
+  "project": {
+    "version": "1.0.0",
+    "restore": {
+      ...
+    },
+    "frameworks": {
+      ...
+    }
+  }
+}
+
+```
+
+### Technical Explanation
+
+The `dotnet nuget why` command will print out the dependency graph of a given package only if it is part of the final graph.
+
+```
+dotnet nuget why [<PROJECT>|<SOLUTION>] <PACKAGE_NAME>
+    [-f|--framework <FRAMEWORK>]
+
+dotnet nuget why -h|--help
+```
+#### Arguments
+
+- PROJECT | SOLUTION
+
+The project or solution file to operate on. If not specified, the command searches the current directory for one. If more than one solution is found, an error is thrown. If more than one project is found, the dependency graph for each project is printed. 
+
+- PACKAGE_NAME
+
+The exact package name/id for which the dependency graph has to be identified. Package name wildcards are not supported.
```

Comment:
Should we also accept multiple packages or a regular expression pattern to match package names?

Most of the existing `dotnet nuget` commands accepts multiple packages that includes a regular expression pattern. For example, [dotnet nuget verify](https://docs.microsoft.com/dotnet/core/tools/dotnet-nuget-verify#arguments) command.

Reply:
I think this is a great idea! Given the limited time of my internship I don't think I will be able to implement it, but I will definitely document it in the spec under additional improvements.

Reply:
I forgot, but this is actually already documented here: https://github.com/NuGet/Home/pull/11875/files#diff-79e5d3f65439f3d176fb83b648e933018d283daf2bb4b4fef783785b4ed79249:~:text=Allow%20the%20customer%20to%20look%20up%20transitive%20dependencies%20of%20more%20than%20one%20package.%20For%20example%3A%20%60dotnet%20nuget%20why%20%5B%3CPROJECT%3E%7C%3CSOLUTION%3E%5D%20packages%20package1%2C%20package2%60%20or%20%60dotnet%20nuget%20why%20%5B%3CPROJECT%3E%7C%3CSOLUTION%3E%5D%20package%20%27nuget.*%27%60.